src/python/pants/backend/shell/shunit2_test_runner.py

```diff
@@ -0,0 +1,244 @@
+    shell_setup: ShellSetup,
+    test_subsystem: TestSubsystem,
+    test_extra_env: TestExtraEnv,
+    global_options: GlobalOptions,
+) -> TestSetup:
+    shunit2_download_file = DownloadFile(
+        "https://raw.githubusercontent.com/kward/shunit2/b9102bb763cc603b3115ed30a5648bf950548097/shunit2",
+        FileDigest("1f11477b7948150d1ca50cdd41d89be4ed2acd137e26d2e0fe23966d0e272cc5", 40987),
+    )
+    shunit2_script, transitive_targets, env = await MultiGet(
+        Get(Digest, DownloadFile, shunit2_download_file),
+        Get(TransitiveTargets, TransitiveTargetsRequest([request.field_set.address])),
+        Get(Environment, EnvironmentRequest(["PATH"])),
+    )
+
+    dependencies_source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            (tgt.get(Sources) for tgt in transitive_targets.dependencies),
+            for_sources_types=(ShellSources, FilesSources, ResourcesSources),
+            enable_codegen=True,
+        ),
+    )
+    dependencies_source_files, field_set_sources = await MultiGet(
+        dependencies_source_files_request,
+        Get(SourceFiles, SourceFilesRequest([request.field_set.sources])),
+    )
+
+    field_set_digest_content = await Get(DigestContents, Digest, field_set_sources.snapshot.digest)
+    # Because a FieldSet corresponds to a file address, there should be exactly 1 file in the
+    # sources. This assumption allows us to simplify determining which shell to use via inspecting
+    # the shebang.
+    if len(field_set_digest_content) != 1:
+        raise AssertionError(
+            f"The file address {request.field_set.address} had sources != 1, which is unexpected: "
+            f"{field_set_sources.snapshot.files}. Please file a bug at "
+            "https://github.com/pantsbuild/pants/issues/new with this error message copied."
+        )
+    original_test_file_content = field_set_digest_content[0]
+
+    validate_source_shunit2_not_included(original_test_file_content, request.field_set.address)
+    updated_test_file_content = FileContent(
+        original_test_file_content.path,
+        # NB: We always run tests from the build root, so we source `shunit2` relative to there.
+        original_test_file_content.content + "\nsource ./shunit2\n".encode(),
+    )
+
+    updated_test_digest, runner = await MultiGet(
+        Get(Digest, CreateDigest([updated_test_file_content])),
+        Get(
+            Shunit2Runner,
+            Shunit2RunnerRequest(
+                request.field_set.address, original_test_file_content, request.field_set.shell
+            ),
+        ),
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            [shunit2_script, updated_test_digest, dependencies_source_files.snapshot.digest]
+        ),
+    )
+
+    env_dict = {
+        "PATH": create_path_env_var(shell_setup.executable_search_path(env)),
+        "SHUNIT_COLOR": "always" if global_options.options.colors else "none",
```

Comment:
Is this a regression? IOW does shunit autodetect based on the presence of a terminal if left to its own devices? I assume this is widely pre-existing, but reflecting the status of the Pants session controlling terminal into Process executions would be part of this puzzle. I don't think we attempt that.

Reply:
Shunit does autodetect but it wasn't working properly. I was hacking on this late at night, but iirc in the Pants integration test it was trying to use colors still.

Further, I think it's a nice service for Pants to unify color handling - if you set `--no-colors`, Pants will make a best-effort attempt to ensure all tools don't use it.

FYI we do this same thing with Pytest:

https://github.com/pantsbuild/pants/blob/f481efed4596b5371828337d58a8de77514b684d/src/python/pants/backend/python/goals/pytest_runner.py#L234